rama-http/src/headers/x_robots_tag_components/robots_tag_components/parser.rs

```diff
@@ -0,0 +1,151 @@
+use crate::headers::x_robots_tag_components::robots_tag_components::Builder;
+use crate::headers::x_robots_tag_components::RobotsTag;
+use http::HeaderValue;
+use rama_core::error::OpaqueError;
+
+pub(crate) struct Parser<'a> {
+    remaining: Option<&'a str>,
+}
+
+impl<'a> Parser<'a> {
+    pub(crate) fn new(remaining: &'a str) -> Self {
+        let remaining = match remaining.trim() {
+            "" => None,
+            text => Some(text),
+        };
+
+        Self { remaining }
+    }
+}
+
+impl<'a> Iterator for Parser<'_> {
+    type Item = Result<RobotsTag, OpaqueError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut remaining = self.remaining?.trim();
+        if remaining.is_empty() {
+            return None;
+        }
+
+        let mut builder = Self::parse_first(&mut remaining).ok()?;
+
+        while let Some((field, rest)) = remaining.split_once(',') {
+            match builder.add_field(field) {
+                Ok(_) => {
+                    remaining = rest;
+                }
+                Err(e) if e.is::<headers::Error>() => {
+                    self.remaining = Some(remaining.trim());
+                    return Some(Ok(builder.build()));
+                }
+                Err(e) => return Some(Err(e)),
+            }
+        }
+
+        match builder.add_field(remaining) {
+            Ok(_) => {
+                self.remaining = None;
+                Some(Ok(builder.build()))
+            }
+            Err(e) if e.is::<headers::Error>() => {
+                self.remaining = Some(remaining.trim());
+                Some(Ok(builder.build()))
+            }
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+impl Parser<'_> {
+    fn parse_first(remaining: &mut &str) -> Result<Builder<RobotsTag>, OpaqueError> {
+        match remaining.find(&[':', ',']) {
+            Some(index) if remaining.chars().nth(index) == Some(',') => {
```

Comment:
you are using high level constructs here instead of `remaining[index]`, but below you anyway do `&remaining[index + 1..];`, might as well do it here already

I also think that that header is only ever expected to containing visible ASCII chars, you might wanna validate that first, somewhere prior to parsing, so that you can safely assume indexing and slicing in parse functions like this